### PR TITLE
Task/t119777 scan

### PR DESCRIPTION
**Pull request by dikaso**
This PR includes:
 - new option to provide plugin to slave
    - this is only for internal usage (for now). If we see broader usage we can standardize interface
 - plugin can hook up on init and on load by defining `setUpOnce` and `onLoad` methods
 - new barcode scan plugin
    - it supports scan with prefix or without him
    - it emits `Slave.ScanFinished` when scan is completed
 
 After this PR is approved I need to:
  - document new plugin
  
 Master will send config on load and plugin will pick up this config:
 
 ```json
{
  "barcodeScan": {
    "treshold": 100,
    "regex": {
      "pattern": "^[A-Za-z0-9.\\-%]$",
      "flags": "i"
    },
    "prefix": "space"
  }
}
 ```
  
  
  Example:
  
  ```js
import slave from '@nsoft/seven-gravity-gateway/slave';
import BarcodeScanPlugin from '@nsoft/seven-gravity-gateway/plugin-barcode-scan';

slave({
   ...,
   plugins: [new BarcodeScanPlugin()],
   ...
})

  ```

### CODE REVIEW

**Review by dev387 — APPROVED** [review]

**Review by Lanchi — APPROVED** [review]

File: test/plugin-barcode-scan.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            document.dispatchEvent(event);
+        });
+      
+        assert(spy.calledWith, {
+            action: 'Slave.ScanFinished',
+            data: {
+                code: result
+            }
+        });
+    });
```
Comment (ZvonimirBusic) [inline test/plugin-barcode-scan.js:57]: Tests failing

```
1) Barcode scan should emit Slave.ScanFinished when scan is finished:
     TypeError: Should wrap property of object
      at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:24:15)
      at Function.spy (node_modules/sinon/lib/sinon/spy.js:45:16)
      at Sandbox.spy (node_modules/sinon/lib/sinon/sandbox.js:275:26)
      at Context.<anonymous> (test/plugin-barcode-scan.js:26:25)
      at processImmediate (internal/timers.js:461:21)

  2) Barcode scan should emit Slave.ScanFinished when scan is finished without prefix:
     TypeError: Should wrap property of object
      at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:24:15)
      at Function.spy (node_modules/sinon/lib/sinon/spy.js:45:16)
      at Sandbox.spy (node_modules/sinon/lib/sinon/sandbox.js:275:26)
      at Context.<anonymous> (test/plugin-barcode-scan.js:44:25)
      at processImmediate (internal/timers.js:461:21)
```

**Review by ZvonimirBusic — CHANGES_REQUESTED** [review]

File: test/plugin-barcode-scan.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    BarcodeScan = require('../plugin-barcode-scan'),
+    barcodeScanPplugin = new BarcodeScan();
+
+
+describe('Barcode scan', function() {
+    var slaveInstance;
+
+    beforeEach(function () {
+        slaveInstance = Gateway();
+    });
```
Comment (ZvonimirBusic) [inline test/plugin-barcode-scan.js:13]: eslint complains about `beforeEach` not being recognized so `"mocha": true` should be included in`.eslintrc` `env` property

File: src/slave_gateway.js [hunk 3]
```diff
@@ -61,6 +61,12 @@ var slaveGateway = {
         if(this.config.worker) {
             this.setWorker();
         }
+
+        if(this.config.plugins && this.config.plugins.length) {
+            this.config.plugins.forEach(function(plugin) {
+                plugin.setUpOnce(this);
```
[the remark this answers is not in the document]
Comment (dikaso) [reply src/slave_gateway.js:67]: Storage plugin cannot be passed to `plugins` array. Storage plugin is passed to `worker.plugins` array. See https://github.com/nsftx/seven-gravity-gateway/wiki/3.-Usage#351-plugins.

Although I agree with you that we should align implementations but I find it out of scope for this PR.

**Review by aida2709 — APPROVED** [review]

File: src/plugin/barcode-scan.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    // Too much difference between characters - which means that this is not scan mode.
+    // First time, difference is equal to current time, so we will extract that from check.
+    // We also prevent scan trigger if somone holds key (e.repeat).
+    // Last flag is when space is trigered as first char, we want to enter in scan mode 
+    if ((difference > config.treshold 
+        && difference !== currentTime 
+        && !isPrefixTriggered)
+        || e.repeat) {
+        codeLoaded = '';
+        previousEventReceived = Date.now();
```
[the remark this answers is not in the document]
Comment (dikaso) [reply src/plugin/barcode-scan.js:35]: Almost, first time we use it in this line `difference = currentTime - previousEventReceived;` so it has to be initiated with "some" value.

[the remark this answers is not in the document]
Comment (ZvonimirBusic) [reply src/plugin/barcode-scan.js:35]: @dikaso Either I misunderstood something or we're not referring to same thing.

Here is code sequence:

1. `currentTime = Date.now();`
2. `previousEventReceived= currentTime;`
3. `previousEventReceived = Date.now();`

Why is no 3. done, it already is Date.now() due to 1 and 2.

Another thing that i realized is that `Date.now()` is called 2 times and will therefore be slightly different. It looks to me that variables to which they are assigned should use same value (`currentTime` and `previousKey.receivedAt`)

[the remark this answers is not in the document]
Comment (ZvonimirBusic) [reply src/plugin/barcode-scan.js]: @dikaso previousEventReceived= currentTime; is still done twice?

**Review by ZvonimirBusic — CHANGES_REQUESTED** [review]
https://github.com/nsftx/seven-gravity-gateway/pull/23#discussion_r781199234

**Review by ZvonimirBusic — APPROVED** [review]

**Pull request merged**